File: app/src/main/java/com/moufee/purduemenus/ui/menu/MenuItemListFragment.java
package com.moufee.purduemenus.ui.menu;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.moufee.purduemenus.R;
import com.moufee.purduemenus.menus.FullDayMenu;
import com.moufee.purduemenus.menus.MenuItem;
import com.moufee.purduemenus.repository.FavoritesRepository;
import com.moufee.purduemenus.util.Resource;

import java.util.ArrayList;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;

/**
 * A Fragment that contains a list of menu items for one Meal at one Dining Court
 * <p/>
 */
public class MenuItemListFragment extends Fragment implements OnToggleFavoriteListener {

    private static final String ARG_DINING_COURT_INDEX = "dining-court-index";
    private static final String ARG_MEAL_INDEX = "meal-index";
    private static final String TAG = "MENU_ITEM_LIST_FRAGMENT";

    // TODO: restructure so that all data stays in ViewModel (data binding?)
    String mDiningCourtName;
    private int mMealIndex = 0;
    private RecyclerView mMenuItemRecyclerView;
    private TextView mNotServingTextView;
    private MenuRecyclerViewAdapter mDataBoundAdapter;
    private DailyMenuViewModel mViewModel;
    @Inject
    ViewModelProvider.Factory mViewModelFactory;
    @Inject
    FavoritesRepository mFavoritesRepository;


    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public MenuItemListFragment() {
    }

    // TODO: Customize parameter initialization
    @SuppressWarnings("unused")
    public static MenuItemListFragment newInstance(String diningCourtName, int mealIndex) {
        MenuItemListFragment fragment = new MenuItemListFragment();
        Bundle args = new Bundle();
        args.putString(ARG_DINING_COURT_INDEX, diningCourtName);
        args.putInt(ARG_MEAL_INDEX, mealIndex);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mDataBoundAdapter = new MenuRecyclerViewAdapter(this);

        if (getArguments() != null) {
            mDiningCourtName = getArguments().getString(ARG_DINING_COURT_INDEX);
            mMealIndex = getArguments().getInt(ARG_MEAL_INDEX);
        }
    }

    @Override
    public boolean toggleFavorite(MenuItem item) {
        if (mViewModel.getFavoriteSet().getValue() == null) return true;
        Log.d(TAG, "toggleFavorite: Adding Favorite " + item.getName() + " " + item.getId());
        if (!mViewModel.getFavoriteSet().getValue().contains(item.getId()))
            mFavoritesRepository.addFavorite(item);
        else
            mFavoritesRepository.removeFavorite(item);
        return true;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_menuitem_list, container, false);


        // Set the adapter
        Context context = view.getContext();
        mMenuItemRecyclerView = view.findViewById(R.id.menu_item_recyclerview);
        mNotServingTextView = view.findViewById(R.id.not_serving_textview);
        mMenuItemRecyclerView.setAdapter(mDataBoundAdapter);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(context);
        if (context.getResources().getConfiguration().screenWidthDp > 500) {
            GridLayoutManager gridLayoutManager = new GridLayoutManager(context, 2);
            layoutManager = gridLayoutManager;
            gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int position) {
                    return mDataBoundAdapter.isHeader(position) ? 2 : 1;
                }
            });
        }
        mMenuItemRecyclerView.setLayoutManager(layoutManager);

        return view;
    }


    private void setListener() {
        mViewModel.getFullMenu().observe(this, fullDayMenuResource -> {
            try {
                if (fullDayMenuResource != null && fullDayMenuResource.data != null && fullDayMenuResource.data.getMenu(mDiningCourtName).isServing(mMealIndex)) {
                    mDataBoundAdapter.setStations(fullDayMenuResource.data.getMenu(mDiningCourtName).getMeal(mMealIndex).getStations());
                    mMenuItemRecyclerView.setVisibility(View.VISIBLE);
                    mNotServingTextView.setVisibility(View.GONE);
                } else {
                    mMenuItemRecyclerView.setVisibility(View.GONE);
                    mNotServingTextView.setText(fullDayMenuResource.data.getMenu(mDiningCourtName).getMeal(mMealIndex).getStatus());
                    mNotServingTextView.setVisibility(View.VISIBLE);
                }
            } catch (Exception e) {
                // this is called frequently due to getStatus() being called on null meals
                // todo: better handling of null meals (Kotlin?)
                mDataBoundAdapter.setStations(new ArrayList<>());
                mMenuItemRecyclerView.setVisibility(View.GONE);
                mNotServingTextView.setText(R.string.not_serving);
                mNotServingTextView.setVisibility(View.VISIBLE);
            }

        });
        // todo: refactor this and above method to remove duplicated code
        mViewModel.getSelectedMealIndex().observe(this, integer -> {
            if (integer == null)
                return;
            mMealIndex = integer;
            Resource<FullDayMenu> fullDayMenuResource = mViewModel.getFullMenu().getValue();
            try {
                if (fullDayMenuResource != null && fullDayMenuResource.data != null && fullDayMenuResource.data.getMenu(mDiningCourtName).isServing(mMealIndex)) {
//                        DiningCourtMenu.Meal meal = fullDayMenuResource.data.getMenu(mDiningCourtIndex).getMeal(mMealIndex);
                    mDataBoundAdapter.setStations(mViewModel.getFullMenu().getValue().data.getMenu(mDiningCourtName).getMeal(mMealIndex).getStations());
                    mMenuItemRecyclerView.setVisibility(View.VISIBLE);
                    mNotServingTextView.setVisibility(View.GONE);
                } else {
                    mMenuItemRecyclerView.setVisibility(View.GONE);
                    mNotServingTextView.setText(fullDayMenuResource.data.getMenu(mDiningCourtName).getMeal(mMealIndex).getStatus());
                    mNotServingTextView.setVisibility(View.VISIBLE);
                }
            } catch (Exception e) {
                mDataBoundAdapter.setStations(new ArrayList<>());
                mMenuItemRecyclerView.setVisibility(View.GONE);
                mNotServingTextView.setText(R.string.not_serving);
                mNotServingTextView.setVisibility(View.VISIBLE);
            }
        });
        mViewModel.getFavoriteSet().observe(this, favoriteIDs -> mDataBoundAdapter.setFavoriteSet(favoriteIDs));
    }


    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
        if (context instanceof AppCompatActivity) {
            mViewModel = ViewModelProviders.of((AppCompatActivity) context, mViewModelFactory).get(DailyMenuViewModel.class);
            setListener();
        }

    }

    @Override
    public void onDetach() {
        super.onDetach();
    }
}

File: app/src/main/java/com/moufee/purduemenus/ui/menu/MenuItemListFragment.kt
package com.moufee.purduemenus.ui.menu

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.moufee.purduemenus.R
import com.moufee.purduemenus.databinding.FragmentMenuitemListBinding
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

/**
 * A Fragment that contains a list of menu items for one meal at one Dining Court
 *
 *
 */
@AndroidEntryPoint
class MenuItemListFragment : Fragment() {

    // TODO: restructure so that all data stays in ViewModel (data binding?)
    lateinit var mDiningCourtName: String
    private lateinit var mViewModel: MenuViewModel
    private lateinit var binding: FragmentMenuitemListBinding
    private lateinit var itemController: MenuItemController

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            mDiningCourtName = requireArguments().getString(ARG_DINING_COURT_NAME, "")
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        binding = FragmentMenuitemListBinding.inflate(inflater)
        binding.lifecycleOwner = viewLifecycleOwner
        binding.viewModel = mViewModel
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        itemController = MenuItemController(object : MenuItemController.AdapterCallbacks {
            override fun onItemLongPressed(item: MenuItemViewObject): Boolean {
                mViewModel.toggleFavorite(item.menuItem)
                return true
            }
        })
        binding.menuItemRecyclerView.adapter = itemController.adapter
        binding.menuItemRecyclerView.itemAnimator = null
        val layoutManager: RecyclerView.LayoutManager =
            if (context?.resources?.configuration?.screenWidthDp ?: 0 > 500) {
                GridLayoutManager(context, 2).apply {
                    itemController.spanCount = 2
                    spanSizeLookup = itemController.spanSizeLookup
                }
            } else LinearLayoutManager(context)
        binding.menuItemRecyclerView.layoutManager = layoutManager
        setListener()
    }

    private fun setListener() {
        lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                mViewModel.getMenuDetailsUiState(mDiningCourtName).collect { state ->
                    itemController.setData(state.items)
                    binding.dataAvailable = state.items.isNotEmpty()
                    binding.notServingTextview.text =
                        if (state.status != null && state.status != "Open") state.status else getString(R.string.no_data)
                    binding.servingTimeTextView.text = state.servingTimeText
                }
            }
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is AppCompatActivity) {
            mViewModel = ViewModelProvider(context).get(MenuViewModel::class.java)
        }
    }

    companion object {
        private const val ARG_DINING_COURT_NAME = "dining-court-name"
        fun newInstance(diningCourtName: String): MenuItemListFragment {
            val fragment = MenuItemListFragment()
            val args = Bundle()
            args.putString(ARG_DINING_COURT_NAME, diningCourtName)
            fragment.arguments = args
            return fragment
        }
    }
}